### PR TITLE
Download upstream docs for dynamically bridged provider

### LINKED ISSUE #2607: Generate Any TF Provider schemas with full examples
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

Currently, schemas generated from `pulumi-terraform-provider` are not aware of their upstream git repo, and use only in-memory docs. When generating docs for the registry, we should run the full documentation pipeline used when generating docs for our other providers. We can do this by setting a flag on our `get-schema` call:

```console
$ pulumi package get-schema terraform-provider $PROVIDER_NAME --docs=full --upstream-repo=$UPSTREAM_REPO
```

The provider would need to download `$UPSTREAM_REPO`'s `docs` or `website` folder and then point docs generation at it.

### PR DESCRIPTION
**Pull request by guineveresaenger**
This pull request adds logic that enables us to generate resource docs for a dynamically bridged provider using the upstream dependency.

Usage for remote providers:
`pulumi package get-schema terraform-provider <registry address> <version> fullDocs=<true|false>`. 

Adds a parameterized arg field  to `pulumi package get-schema terraform-provider <registry address> <version>` called `fullDocs`, which when set to `true` will instruct the bridge to `git clone --depth 1 -b <version> <terraform provider github repo> <local dir for dynamic docs>`. This allows us to keep the exact same docs logic we have established. The shallow clone targets the exact doc version we need. (*)

For this, we infer the github repo from the OpenTofu org name. It is based on the assumption that `registry.opentofu.org/org/foo` is based on a provider that lives at `github.com/org/terraform-provider-foo`. OpenTofu says their protocol follows that of the HashiCorp Terraform Registry which [requires an org/user and a provider name of the format `terraform-provider-foo`](https://developer.hashicorp.com/terraform/registry/providers/publishing) and we have historical evidence that GitHub is the most commonly used source host for terraform providers. See also https://github.com/opentofu/registry/issues/1337.

Usage for local providers:
`pulumi package get-schema terraform-provider <local-path> upstreamRepoPath=<localPath>`. Here, the local docs path gets read directly into `upstreamRepoPath`. 

The one thing that is perhaps missing here is to expand the remote args to allow to take a source repo as an additional argument. However, our internal use case is primarily one of automation, so this is not an immediate need.

(*)Cloning docs for the AWS terraform provider at v5.70.0 has the following time performance:
```
git clone --depth 1 -b v5.70.0  awsDir  
2.06s user 
1.62s system 
34% cpu 
10.713 total
```

Fixes #2607

### CODE REVIEW

File: dynamic/main.go
Comment (iwahbe) [inline dynamic/main.go:81]: Instead of keeping track of both `schemaDocsOnly` and `fullDocs` (and keeping them in sync), we can simplify to one variable.
```suggestion
				XInMemoryDocs:            !fullDocs,
				XLoadUpstreamRepoForDocs: fullDocs,
```

File: dynamic/parameterize/args.go
Comment (iwahbe) [inline dynamic/parameterize/args.go:75]: By convention, the `[]` mean optional.
```suggestion
		return Args{}, fmt.Errorf("expected to be parameterized by 1-3 arguments: <name> [version] [fullDocs]")
```

File: dynamic/info.go
Comment (guineveresaenger) [reply dynamic/info.go:37]: hence the TODO 😝

Comment (iwahbe) [reply dynamic/info.go:37]: I'm never sure what TODOs are meant to stay in with an issue after merge and which ones are meant as a "must fix before merging".

File: pkg/tfgen/generate.go
Comment (iwahbe) [inline pkg/tfgen/generate.go:947]: Let's not have extra changes here.
```suggestion
```

File: pkg/tfbridge/info/info.go
Comment (iwahbe) [inline pkg/tfbridge/info/info.go:133]: Who is this comment for?

File: dynamic/info.go
Comment (iwahbe) [inline dynamic/info.go:1]: We skip all examples on L46. We don't want to do that when we are generating full examples.

File: dynamic/info.go
Comment (iwahbe) [inline dynamic/info.go:100]: We should set `prov.Repository`.
```suggestion
		urlFields := strings.Split(value.Remote.URL, "/")
		prov.GitHubOrg = urlFields[len(urlFields)-2]
		prov.Repository = fmt.Sprintf("https://github.com/%s/terraform-provider-%s",
			prov.GitHubOrg, urlFields[len(urlFields)-1])
```

File: pkg/tfbridge/info/info.go
Comment (guineveresaenger) [reply pkg/tfbridge/info/info.go:133]: Maintainers. It's set implicitly here, so it makes sense to document this behavior.

File: dynamic/info.go
Comment (guineveresaenger) [reply dynamic/info.go:1]: Thank you for pointing out this wrinkle.

It turns out that's...not what happens, or perhaps I misunderstand.

The examples aren't being _translated_. They're not being skipped; they're in fact still in the schema. (In fact, I'm not entirely sure I understand the purpose/implementation detail of this setting, as setting it to `false` has no effect on the generated schema.)

Comment (guineveresaenger) [reply dynamic/info.go:1]: Update: It turns out that `convertExamples` is never called. The `SkipExamples` field isn't actually ever used in the call stack.
Was this field set to _disable_ behavior? It appears that something isn't hooked up correctly.

Comment (iwahbe) [reply dynamic/info.go:1]: It's a flag I turned on in the hope it would reduce time spent on example generation. If it doesn't do *anything at all*, feel free to remove it.

It was *supposed* to skip converting examples.

File: pkg/tfbridge/info/info.go
Comment (iwahbe) [reply pkg/tfbridge/info/info.go:133]: I wouldn't say its *set implicitly*. It is *set* in `dynamic/info.go`. If you are working in `/dynamic`, then you will see where it is set. If you are implementing a bridged provider, then this is not helpful for you, since it's relevant only to dynamic providers. `info/info` feels like the wrong place for this comment.

This feels spiritually equivalent to annotating a function with where it is called, which we don't and shouldn't do.

File: dynamic/main.go
Comment (iwahbe) [inline dynamic/main.go:160]: Instead of calling this `versionWithPrefix`, let's call it what we assume it to be: the tag at which we are downloading the provider.
```suggestion
					tag := "v" + info.Version
```

File: dynamic/parameterize/args.go
Comment (iwahbe) [inline dynamic/parameterize/args.go:34]: What does "(possibly empty)" mean in this context? I'm also not sure why this is an "argument"

```suggestion
	// Docs indicates if full schema documentation should be generated.
```

File: dynamic/parameterize/args.go
Comment (iwahbe) [inline dynamic/parameterize/args.go:43]: Maybe we should call this `UpstreamRepoPath`, since that's what it is. 

In any case, its not the path to docs, but to the repo that contains the docs.

```suggestion
	// UpstreamRepoPath (if provided) is the local path to the dynamically bridged Terraform provider's repo.
	//
	// If set, full documentation will be generated for the provider. If not set, only documentation from the TF provider's schema will be used.
	UpstreamRepoPath string
```

File: dynamic/parameterize/args.go
Comment (iwahbe) [inline dynamic/parameterize/args.go:48]:
```suggestion
```

File: dynamic/parameterize/args.go
Comment (iwahbe) [inline dynamic/parameterize/args.go:55]: Let's validate that `docsLocation` exists here.

Comment (iwahbe) [inline dynamic/parameterize/args.go:55]: If we are going to require a prefix of `docsLocation=`, it needs to be documented.
```suggestion
				return Args{}, fmt.Errorf("path based providers are only parameterized by 2 arguments: <path> [docsLocation=<localPath>]")
```

File: dynamic/testdata/TestSchemaGenerationFullDocs/hashicorp/random-3.6.3.golden
Comment (iwahbe) [inline dynamic/testdata/TestSchemaGenerationFullDocs/hashicorp/random-3.6.3.golden:226]: It looks like we are not converting examples, which we definitely need to do.

File: dynamic/info.go
Comment (guineveresaenger) [reply dynamic/info.go:1]: Update: It turns out this flag _does_ skip example conversion! But we weren't calling into that code at all, because we weren't calling tfgen.Generate().

So we can use it here, but only if we call the full Generate() process.

File: dynamic/info.go
Comment (guineveresaenger) [reply dynamic/info.go:100]: I wonder if there's a better way to differentiate between "this provider has pulumi source code" and "this provider only has terraform source code".
While the [docs are pretty clear that the field refers to the source code repo](https://github.com/pulumi/pulumi-terraform-bridge/blob/master/pkg/tfbridge/info/info.go#L70), for Pulumi-maintained providers this field is expected to be a Pulumi provider repository as opposed to a Terraform one.
And elsewhere does [express this difference](https://github.com/pulumi/pulumi-terraform-bridge/blob/2a8b877dbe500f33980524c68855f0e732c3d6aa/pkg/tfgen/generate_schema.go#L588).

**Comment by guineveresaenger** [comment]
This may not be the way we want to do this, but I changed the function signature of `Generate()` and `UnstableGenerateFromSchema` to return a package schema so that we can access it from an outside library.

Alternatives:
- Implementing a separate `MakeSchemaWithExamples(generator)` function in `generate.go` that takes a pre-created Generator as an argument and returns a package schema for consumption by `dynamic`
- Generating the "lightweight" schema first, as before, and then making a separate call to `UnstableGenerateFromSchema`.

Unfortunately, adding extra options to the GeneratorOpts only gets us halfway - we must call Generate() or its parts somewhere in `dynamic`.

I've gone with changing the function signature to avoid writing yet another generator function with a narrowly scoped purpose, but am not married to the idea.

File: Makefile
Comment (guineveresaenger) [reply Makefile:9]: We need this bump, otherwise the test won't pass.

File: dynamic/provider_test.go
Comment (guineveresaenger) [reply dynamic/provider_test.go:472]: > Why are we defining a local struct only to instantiate one copy?
- Test information at a glance
- More easily add subtests in the future if needed
- Allows us to run individual tests by name when using an integrated IDE test framework
- Makes the name of subtest far more obvious for running it in the terminal

Is there a downside here?

File: dynamic/info.go
Comment (guineveresaenger) [reply dynamic/info.go:46]: We never go to example generation in the normal case. But point taken on the default.

File: dynamic/parameterize/args.go
Comment (iwahbe) [inline dynamic/parameterize/args.go:105]: Let's be consistent with `upstreamRepoPath` for local providers.

```suggestion
		return Args{}, fmt.Errorf("expected to be parameterized by 1-3 arguments: <name> [version] [fullDocs=<true|false>]")
```

**Review by iwahbe — APPROVED** [review]

File: dynamic/provider_test.go
Comment (iwahbe) [reply dynamic/provider_test.go]: Verbosity and the chance of rot. It's an abstraction we aren't using.

File: dynamic/testdata/TestSchemaGeneration/Backblaze/b2-0.8.9.golden
Comment (iwahbe) [inline dynamic/testdata/TestSchemaGeneration/Backblaze/b2-0.8.9.golden:3]: This is a regression.

**Pull request merged**
